### PR TITLE
Fix copilot bugs

### PR DESCRIPTION
**Pull request by rushatgabhane**
### Fixed Issues

$ https://github.com/Expensify/App/issues/48261
https://github.com/Expensify/App/issues/48246
https://github.com/Expensify/App/issues/48252
https://github.com/Expensify/App/issues/48253
https://github.com/Expensify/App/issues/48259
https://github.com/Expensify/App/issues/48244

#### New issues @mkhutornyi  -

https://github.com/Expensify/App/issues/48965
https://github.com/Expensify/App/issues/48966
https://github.com/Expensify/App/issues/48967
https://github.com/Expensify/App/issues/48972

PROPOSAL:

### Tests

Pre-req: be on copilot beta

### Long name 
1. Go to profile
2. Set your name as a very long name
3. Go to settings page
4. Verify that there is a space between your name and status emoji

### Subscription page

1. Go to Account settings.
2. Go to Subscription.
3. Click Account switcher.
4. Select another account.
6. Verify that the subscription page has a loading indicator.

### Avatar error

1. Go to Account settings > Profile.
2. Upload a large image that will trigger error.
3. Click on the account switcher.
4. Select another account.
6. Verify that the invalid avatar format error will disappear after switching account.

### Interactivity
1. Go to account with no copilot
2. Go to settings page
3. Click you account name
4. Verify that clicking name does not show interactivity 

### Avatar skeleton
1. GO to profile page
2. Switch accounts
3. Verify that avatar in profile page shows skeleton

### Selected access level
1. Go to Account settings > Security.
2. Click Add copilot.
3. Select a user.
4. Select an access level.
5. Click Access level on the confirmation page
6. Verify that the selected access level will be highlighted.

### Fallback avatar 
1. Go to Account settings > Security.
2. Click on the account switcher.
3. Switch to another acccount.
4. After everything loads, switcher to the original account while staying on Security page.
5. Scroll down the Security page.
6. Verify that the account under "You can access these accounts via the account switcher:" should have an avatar.

### Add yourself as copilot 

1. Go to settings > security > add Copilot
2. Write the email you used to login
3. Verify that you see "No results found message"

### Learn more link

1. Settings > Security
2. Reduce the screen width to mobile size
3. Verify that the "learn more" link has correct styles

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/5a0866d6-4028-4e89-830d-1ab897c919b6

image

https://github.com/user-attachments/assets/17be3818-8f8d-4912-8504-b1378b54a1fb

https://github.com/user-attachments/assets/dca55b0e-ce6a-4df0-81aa-f63e651a6d47

image

https://github.com/user-attachments/assets/4d286bec-82aa-4bb3-85c4-0cbb53ff7377

https://github.com/Expensify/App/issues/48965 -

image

https://github.com/Expensify/App/issues/48966 -  

image

https://github.com/Expensify/App/issues/48967 -

image

https://github.com/Expensify/App/issues/48972 -

image

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by dangrous** [comment]
I know this is a draft but wanted to call out two bugs not yet listed in the list in the PR description:

* https://github.com/Expensify/App/issues/48244
* [IOS: BUG: While offline, Alert does not show up on changing accounts.](https://github.com/Expensify/App/pull/47338#issuecomment-2313666208)

**Comment by rushatgabhane** [comment]
added them to the list

**Comment by rushatgabhane** [comment]
- [x] https://github.com/Expensify/App/issues/48261
- [ ] https://github.com/Expensify/App/issues/48245
- [ ] https://github.com/Expensify/App/issues/48246
- [ ] https://github.com/Expensify/App/issues/48252
- [ ] https://github.com/Expensify/App/issues/48253
- [ ] https://github.com/Expensify/App/issues/48259
- [ ] https://github.com/Expensify/App/issues/48292
- [ ] https://github.com/Expensify/App/pull/47338#issuecomment-2313666208
- [ ] https://github.com/Expensify/App/issues/48244

**Comment by rushatgabhane** [comment]
cc: @dangrous

**Comment by rushatgabhane** [comment]
> https://github.com/Expensify/App/pull/47338#issuecomment-2313666208

@dangrous this was fixed in the PR itself. So I've removed it from the list


https://github.com/user-attachments/assets/a400d3ce-e56b-4c49-91fc-db3517e66ca8

**Comment by rushatgabhane** [comment]
All bugs here fixed - https://github.com/Expensify/App/pull/48517#issuecomment-2336781614

cc @mkhutornyi

**Comment by rushatgabhane** [comment]
> Bug: app crashes while switching account. And blank account info

unrelated to copilot actually. and being fixed in another issue

File: src/components/AccountSwitcher.tsx
Comment (rushatgabhane) [reply src/components/AccountSwitcher.tsx:145]: @mkhutornyi you can look at the commits for that - https://github.com/Expensify/App/pull/48517/commits

File: src/components/AvatarWithImagePicker.tsx
Comment (rushatgabhane) [reply src/components/AvatarWithImagePicker.tsx:379]: nice catch!

Comment (rushatgabhane) [reply src/components/AvatarWithImagePicker.tsx:379]: i think i'll just override the style then

**Comment by dangrous** [comment]
@rushatgabhane are we going to be able to add the new copilot bugs into this PR as well? I believe you're assigned to all of them currently. Thanks!

**Comment by rushatgabhane** [comment]
yes will do

**Comment by dangrous** [comment]
Do we think this will be ready to review on Monday? I'd love to be able to get this merged then and then see if any new bugs crop up.

File: src/components/AvatarWithImagePicker.tsx
Comment (rushatgabhane) [reply src/components/AvatarWithImagePicker.tsx:379]: image


image

**Comment by rushatgabhane** [comment]
@mkhutornyi for ease of review i added new commits with their issue number. hope that helps.

5 new issues for you to review - 

https://github.com/Expensify/App/issues/48965
https://github.com/Expensify/App/issues/48966
https://github.com/Expensify/App/issues/48967
https://github.com/Expensify/App/issues/48972

ty!

File: src/components/AvatarWithImagePicker.tsx
Comment (rushatgabhane) [reply src/components/AvatarWithImagePicker.tsx]: fixed

**Comment by rushatgabhane** [comment]
@dangrous all done here

File: src/pages/settings/Subscription/SubscriptionSettingsPage.tsx
Comment (rushatgabhane) [reply src/pages/settings/Subscription/SubscriptionSettingsPage.tsx:36]: I don't think it is worth improving this more. The subscription data is simply not in Onyx, so we have to show not found page

Comment (rushatgabhane) [reply src/pages/settings/Subscription/SubscriptionSettingsPage.tsx:36]: this is the conditon now. the most we can do is to show a loader when the app is loading.

```
    if (!subscriptionPlan && isAppLoading) {
        return <FullScreenLoadingIndicator />;
    }
    if (!subscriptionPlan) {
        return <NotFoundPage />;
    }
```

**Comment by rushatgabhane** [comment]
> Is it a bug from our PR ?, the error message is not visible on iOS native:

looks like it let me see

**Review by allgandalf — APPROVED** [review]
## Summary of all the bugs attached to this PR ♻️ :

### Fixed:
- [x] https://github.com/Expensify/App/issues/48261
- [x] https://github.com/Expensify/App/issues/48252
- [x] https://github.com/Expensify/App/issues/48253
- [x] https://github.com/Expensify/App/issues/48259
- [x] https://github.com/Expensify/App/issues/48244 
- [x] https://github.com/Expensify/App/issues/48965 
- [x] https://github.com/Expensify/App/issues/48967
- [x] https://github.com/Expensify/App/issues/48972


### Partially Fixed:

- [ ] https://github.com/Expensify/App/issues/48246 

[Comments](https://github.com/Expensify/App/pull/48517/files#r1761479480) from @rushatgabhane :

> I don't think it is worth improving this more. The subscription data is simply not in Onyx, so we have to show not found page


I agree with this conclusion! There isn't a way that we know of to get the subscription data from `Onyx`


- [ ] https://github.com/Expensify/App/issues/48966 

This bug is more of a `BE` bug, @dangrous you can refer to @rushatgabhane 's comments [here](https://expensify.slack.com/archives/C06ML6X0W9L/p1726505388396179?thread_ts=1726494617.816779&cid=C06ML6X0W9L).


But the above 2 shouldn't block this PR, this PR looks great and fixes a lot of bugs, had fun reviewing this one 🚀 all yours @dangrous

File: src/pages/settings/Subscription/SubscriptionSettingsPage.tsx
Comment (dangrous) [reply src/pages/settings/Subscription/SubscriptionSettingsPage.tsx:36]: Yeah I think that's fine for now

**Comment by dangrous** [comment]
Running through this, it's all looking good so far! Commenting here when I find bugs so you might get a couple messages.

I noticed that we don't differentiate between errors on the new validate code form. I tried adding a copilot from an unvalidated account, and got a `401 Delegator account must be validated` error returned from the server, but all the form said was that it was an incorrect validation code

image

Although now that I'm writing it I'm realizing it might be something we should fix on the backend? Like if the verification code is correct (for all of these new flows), should that count to validate the account? cc @NikkiWines since you wrote the new code on the backend.

**Comment by dangrous** [comment]
For https://github.com/Expensify/App/issues/48246 I think we might want a fallback for if the delegator account doesn't have a subscription but the delegate does, or vice versa - that is, if you're on the subscription page and swap accounts to a user who doesn't have it. Maybe not found is fine here.

**Review by dangrous — CHANGES_REQUESTED** [review]
Fallback icon works for now for https://github.com/Expensify/App/issues/48966, I will look into the back end fix.

Other than those commented bugs, I think we're almost there! Let me know what you think about those, and I agree that we can skip the ones @allgandalf called out

**Comment by rushatgabhane** [comment]
> I noticed that we don't differentiate between errors on the new validate code form. I tried adding a copilot from an unvalidated account, and got a 401 Delegator account must be validated error returned from the server, but all the form said was that it was an incorrect validation code

you're right. we should fix on backend. We have to return onyx data and then frontend can show that.

**Comment by dangrous** [comment]
Ready to merge this if you agree with https://github.com/Expensify/App/pull/48517#issuecomment-2357266910 - should we leave as is? Or would it be better to navigate from subscription page back to profile page or something?

**Comment by rushatgabhane** [comment]
> Ready to merge this if you agree with #48517 (comment) - should we leave as is? Or would it be better to navigate from subscription page back to profile page or something?

we don't have this pattern to fallback to another route. so i think we should do nothing here 👍

**Review by dangrous — APPROVED** [review]
okay great! BE changes needed as noted to fix the remaining bugs but this gets 99% of them and looks great!

**Pull request merged**
